enable deterministic path for index_copy_cuda with index_put

yuguo68 opened the pull request:
Summary: this is similar to index_add_cuda with index_put accumulate = True

Test Plan:
buck test mode/opt //caffe2/test:torch_cuda -- test_index_copy_deterministic

    ✓ ListingSuccess: caffe2/test:torch_cuda - main (9.229)
    ✓ Pass: caffe2/test:torch_cuda - test_index_copy_deterministic_cuda (test_torch.TestTorchDeviceTypeCUDA) (25.750)
    ✓ Pass: caffe2/test:torch_cuda - main (25.750)

Differential Revision: D28291041

kurtamohler commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -507,6 +502,17 @@ Tensor & index_copy_(Tensor & self, int64_t dim, const Tensor & index, const Ten
   TORCH_CHECK_INDEX(source.dim() == 0 || numIndices == source.size(dim),
           "index_copy_(): Number of indices (", numIndices, ") should be equal to source.size(dim) (", source.size(dim), ")");
 
+  if (self.device().type() == DeviceType::CUDA && globalContext().deterministicAlgorithms()){
```
Could you please add the `Note [Enabling Deterministic Operations]` here? https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/Context.h#L124

yuguo68 replied:
thanks, added in the latest version.

kurtamohler replied:
Cool thanks!

yuguo68 committed:
make index_put with accumulate=False deterministic on CPU and CUDA

Differential Revision: D28290699

fbshipit-source-id: 6a19250e77b764e951f3c69f028ce6f6176dbfd4

yuguo68 committed:
enable deterministic path for index_copy_cuda with index_put (#57870)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/57870

this is similar to index_add_cuda with index_put accumulate = True

Test Plan:
buck test mode/opt //caffe2/test:torch_cuda -- test_index_copy_deterministic

    ✓ ListingSuccess: caffe2/test:torch_cuda - main (9.229)
    ✓ Pass: caffe2/test:torch_cuda - test_index_copy_deterministic_cuda (test_torch.TestTorchDeviceTypeCUDA) (25.750)
    ✓ Pass: caffe2/test:torch_cuda - main (25.750)

Reviewed By: ngimel

Differential Revision: D28291041

fbshipit-source-id: acf01003cacfb10bdf43b063fa1f3a653c44e9cc

mruberry commented:
Unlanding. Sorry @yuguo68, this appears to have broken lint:

```
test/test_torch.py:5240:16: error: Incompatible return value type (got "Tuple[Tensor, Tensor, Tensor]", expected "Tuple[Tensor]")  [return-value]
test/test_torch.py:5245:29: error: Need more than 1 value to unpack (3 expected)  [misc]
test/test_torch.py:5247:39: error: Cannot determine type of 'x'  [has-type]
test/test_torch.py:5247:47: error: Cannot determine type of 'index'  [has-type]
test/test_torch.py:5247:54: error: Cannot determine type of 'src'  [has-type]
test/test_torch.py:5249:18: error: Cannot determine type of 'x'  [has-type]
test/test_torch.py:5250:26: error: Cannot determine type of 'index'  [has-type]
test/test_torch.py:5253:47: error: Cannot determine type of 'src'  [has-type]
test/test_torch.py:5255:47: error: Cannot determine type of 'src'  [has-type]
test/test_torch.py:5257:47: error: Cannot determine type of 'src'  [has-type]
test/test_torch.py:5264:29: error: Need more than 1 value to unpack (3 expected)  [misc]
test/test_torch.py:5268:38: error: Cannot determine type of 'x'  [has-type]
test/test_torch.py:5268:46: error: Cannot determine type of 'index'  [has-type]
test/test_torch.py:5268:53: error: Cannot determine type of 'src'  [has-type]
test/test_torch.py:5270:41: error: Cannot determine type of 'x'  [has-type]
test/test_torch.py:5270:49: error: Cannot determine type of 'index'  [has-type]
test/test_torch.py:5270:56: error: Cannot determine type of 'src'  [has-type]
test/test_torch.py:5275:44: error: Cannot determine type of 'x'  [has-type]
test/test_torch.py:5275:52: error: Cannot determine type of 'index'  [has-type]
test/test_torch.py:5275:59: error: Cannot determine type of 'src'  [has-type]
```

I'm not sure why lint didn't run or why it's suddenly interested in typing the test files, however. @samestep, any ideas?

yuguo68 commented:
reland in https://github.com/pytorch/pytorch/pull/58144

yuguo68 commented:
mypy requires Tuple[Tensor, Tensor, Tensor] rather than Tuple[Tensor] for 3 tensors tuple.
